Fixed #30111 -- Fixed AppRegistryNotReady error with django.contrib.postgres in INSTALLED_APPS.

nasirhjafri opened the pull request:
When contrib.postgres is in INSTALLED_APPS it triggers `from django.db.migrations.writer import MigrationWriter` import in https://github.com/django/django/blob/7185ea6902532eb195d0575d1bf1492ca9d45dea/django/contrib/postgres/apps.py#L8 which further triggered https://github.com/django/django/blob/7185ea6902532eb195d0575d1bf1492ca9d45dea/django/db/models/base.py#L99 in ModelBase Meta class.
ModelBase checks for `self.check_apps_ready()` while apps have not loaded completely yet. So an exception is raised `raise AppRegistryNotReady("Apps aren't loaded yet.")`

Fix: I've changed `django/django/contrib/postgres/apps.py` to import MigrationWriter only when needed to avoid `check for apps ready` cycle.

https://code.djangoproject.com/ticket/30111

charettes commented:
#10862 was another PR with a test.

jdufresne commented:
See  #10862 for a potential test case.

My comment from that PR:

I wonder if this is exposing a deeper issue with MigrationWriter.register_serializer(). I believe this means a similar workaround is required for all apps.py files.

nasirhjafri commented:
@jdufresne Thanks, I've added tests cases. 
I've moved the imports in `django/db/migrations/writer.py` instead ref: https://github.com/django/django/pull/10861/commits/dbd8c7fb1316b523afae895f8e8e491df4d2e546

This would fix the issue with `MigrationWriter` at some levels and we won't have to do the similar workaround for all apps.py files that use this feature.

cansarigol commented:
I was trying to move `MigrationRecorder` as your solution. better solution than duplicate code and no need to tackle all apps.py.

nasirhjafri commented:
buildbot, test on oracle.

nasirhjafri commented:
@charettes Can you please review it? I'm not sure why some test builds are failing as tests are working fine locally.

jdufresne commented:
@nasirhjafri We normally open PRs against the "master" branch and the committer will backport it to all appropriate branches. If you're not familiar, please see this article for how to change the base branch:

https://github.blog/2016-08-15-change-the-base-branch-of-a-pull-request/

nasirhjafri commented:
Hi @jdufresne Is there a way to remove commits
```
@carltongibson
[2.2.x] Bumped version for Django 2.2 alpha 1.
ce26b21
 @carltongibson
[2.2.x] Bumped django_next_version in docs config.
``` from the branch?

felixxm commented:
@nasirhjafri You don't need, just rebase your branch from `master`.

carltongibson commented:
Give me a sec, I'll adjust and rebase.

charettes commented on tests/postgres_tests/integration/manage.py:
```diff
@@ -0,0 +1,7 @@
+if __name__ == '__main__':
+    os.environ['DJANGO_SETTINGS_MODULE'] = 'settings'
+    from django.core.management import execute_from_command_line
+    execute_from_command_line(sys.argv)
```
No need for this module.

charettes commented on tests/postgres_tests/test_integration.py:
```diff
@@ -0,0 +1,23 @@
+    def test_management_command(self):
+        manage_py = os.path.join(os.path.dirname(__file__), 'integration', 'manage.py')
+        result = subprocess.run(
+            [sys.executable, manage_py, 'check'],
```
You can call `['python', '-m', 'django', '--settings', 'integration_settings', 'check']` instead to avoid creating a dedicated `manage.py` file.

jdufresne replied:
In the suggested command list, `'python'` should `sys.executable` to ensure the same Python that is running the tests is used in the subprocess.

charettes commented on tests/postgres_tests/integration/settings.py:
```diff
@@ -0,0 +1,5 @@
+
+INSTALLED_APPS = [
+    'django.contrib.postgres',
+]
```
No need to nest, move to ` tests/postgres_tests/integration_settings.py`

charettes commented on django/db/migrations/writer.py:
```diff
@@ -199,6 +198,10 @@ def as_string(self):
+        # Defer import since `MigrationLoader` uses `MigrationRecorder` which
+        # declares `Migration` model.
+        from django.db.migrations.loader import MigrationLoader
+
```
Have we explored making `MigrationRecorder.Migration` to be lazily created on the first class access instead?

nasirhjafri replied:
I think we should move https://github.com/django/django/blob/6713926ebe22172e50f283185f969275c326416d/django/db/migrations/recorder.py#L22 to a new file `django/django/db/migrations/migration.py` and import lazily here https://github.com/django/django/blob/6713926ebe22172e50f283185f969275c326416d/django/db/migrations/recorder.py#L44 and here https://github.com/django/django/blob/6713926ebe22172e50f283185f969275c326416d/django/db/migrations/recorder.py#L40 
What do you think?

carltongibson commented:
Not sure about the failures. They've changed. Initial Windows failure was about failing to  checkout the repo correctly. These are MariaDB/MySQL only, and they seem totally unrelated to the PR (Also there were only two against https://github.com/django/django/commit/14135e267da5a8bf4df655f1708fe228948bfd0a vs three now.

nasirhjafri commented:
Hi @charettes Need your opinion on two possible solutions.
The first one it here https://github.com/django/django/pull/10861/commits/76d78db314ff363b2407fe62c30ccb0dcaf4238f in which we defer the import.
Another one his here https://github.com/django/django/pull/10861/commits/4fe26eee6e55f09f18c1742bd1c24c4e49c7ffce which allows MigrationRecorder.Migration to be lazily created on the first class access instead. 
I think the second one is better. 

cc: @carltongibson @jdufresne

nasirhjafri commented:
Thanks all for reviewing. :)

cansarigol commented:
Should we move Migration model to models.py and change init like this?
```
def __init__(self, connection):
        self.connection = connection
        from .models import Migration
        self.Migration = Migration
```

nasirhjafri commented:
Hi @cansarigol , I don't think that moving `MigrationRecorder.Migration` to models.py and exposing it to the world is a good idea as there is another class named Migration in https://github.com/django/django/blob/master/django/db/migrations/migration.py which is widely used.

cansarigol commented:
I got your objection but I just offer to move `class Migration(models.Model):` not `django.db.migrations.migration.Migration`

nasirhjafri commented:
My objection is regarding `class Migration(models.Model):`. 
Moving it to `models.py` will make it importable and we might need to document the difference between `django.db.migrations.migration.Migration` and `django.db.migrations.models.Migration`. 
Exposing an inner class which is only used in one module, to the whole package and making it importable isn't a good idea. I hope you got my point.

cansarigol commented:
ok you are right thanks for the description

nasirhjafri commented:
Hi @charettes , @jdufresne approved the PR and I've squshed to one commit. Can you please merge it.
Many thanks,

carltongibson commented on django/db/migrations/recorder.py:
```diff
@@ -32,8 +20,24 @@ class Meta:
+    """
     def __init__(self, connection):
         self.connection = connection
+        self.Migration = migration()
```
> Have we explored making MigrationRecorder.Migration to be lazily created on the first class access instead?

Will this have the `Migration` class created each time `MigrationRecorder` is instantiated? Any issues with that? Any preference for making it a static attribute on the class, so it's only created once?

nasirhjafri replied:
@carltongibson No class definitions are only created once when the interpreter reaches in the function `migration`. `self.Migration = migration()` will surely set Migration class again into `self.Migration`, but python interpreter only creates the class definition the first time it encounters that class. 
Moreover, making it static won't fix the issue. If we make it static importing this file would trigger the `migration()` function and we would be back where we started.

nasirhjafri replied:
@carltongibson What do you think if we make it as an attribute?
like
```python
    @property
    def Migration(self):
        class Migration(models.Model):
            app = models.CharField(max_length=255)
            name = models.CharField(max_length=255)
            applied = models.DateTimeField(default=now)

            class Meta:
                apps = Apps()
                app_label = "migrations"
                db_table = "django_migrations"

            def __str__(self):
                return "Migration %s for %s" % (self.name, self.app)

        return Migration
```

carltongibson replied:
Is that correct?

```
>>> def a():
...     class A:
...         pass
...         
...     return A
...
>>> a1 = a()
>>> a2 = a()
>>> a1
__main__.a.<locals>.A
>>> a2
__main__.a.<locals>.A
>>> a1 is a2
False
>>> a1 == a2
False
```

Whereas if `Migration` is a `@property` that lazily creates `Migration`, storing the reference on the class, it'll always be the same. 

I'm not 100% sure it'd raise any problems but, `ModelBase.__new__()` is a little bit hairy, and dynamic model creation isn't really supported, I think for that reason. Hence my query.

carltongibson replied:
@nasirhjafri Our comments crossed. 

Take this class: 

```
class T: 
    
    _B = None
     
    @property 
    def a(self): 
        class A: 
            pass 
    
        return A
    
    @property
    def b(self):
        if self.__class__._B is not None: 
            class B:
                pass
            
            self.__class__._B = B
        return self.__class__.B
```

Behaviour is thus: 

```
>>> T
__main__.T
>>> t = T()
>>> a1 = t.a
>>> a2 = t.a
>>> a1 is a2
False
>>> b1 = T.b
>>> b2 = T.b
>>> b1 is b2
True
```

Your example is like `a()`, whereas `b()` is more like what I took Simon's initial comment to mean. I don't know if it's the best idiom, or if it's 100% necessary, but that's the sort of thing I have in mind.

carltongibson replied:
Clearly using `__class__` isn't what I mean here.

jdufresne replied:
I think @carltongibson's suggestion is a good one. The `Migraiton` class should only ever be created once. Using the `classproperty` decorartor in `django/utils/decorators.py` may make the code simpler.

jdufresne replied:
Sorry for the double post, but just to be explicit, something like:

```py
@classproperty
def migration(cls):
    if cls._migration_class is None:
        class Migration:
            ...

        cls._migration_class = Migration
    return cls._migration_class
```

carltongibson replied:
Thanks @jdufresne. I knew there has to be a nicer way... 🙂

nasirhjafri replied:
Hi @carltongibson, I've improvised my fix. Can you please review it?
Thanks

carltongibson commented on django/db/migrations/recorder.py:
```diff
@@ -18,30 +19,39 @@ class MigrationRecorder:
-            app_label = "migrations"
-            db_table = "django_migrations"
+    @classproperty
+    def migration(cls):
```
And if we just call this `Migration` (with the capital `M`) can we avoid the other changes later in the file?

nasirhjafri replied:
Done!

timgraham commented on django/db/migrations/recorder.py:
```diff
@@ -18,19 +19,26 @@ class MigrationRecorder:
+
+                class Meta:
+                    apps = Apps()
+                    app_label = "migrations"
```
While we're modfying these lines, you could switch all these strings to use single quotes. And perhaps the property deserves a docstring or comment explaining the "why" of the classproperty approach.

timgraham commented on tests/postgres_tests/test_integration.py:
```diff
@@ -0,0 +1,21 @@
+
+
+@unittest.skipIf(psycopg2 is None, 'psycopg2 is required')
+class PostgresIntegrationTest(SimpleTestCase):
```
Test -> Tests (for future expansion)

timgraham commented on tests/postgres_tests/test_integration.py:
```diff
@@ -0,0 +1,21 @@
+
+@unittest.skipIf(psycopg2 is None, 'psycopg2 is required')
+class PostgresIntegrationTest(SimpleTestCase):
+    def test_management_command(self):
```
`test_check`?

timgraham commented on tests/postgres_tests/test_integration.py:
```diff
@@ -0,0 +1,21 @@
+class PostgresIntegrationTest(SimpleTestCase):
+    def test_management_command(self):
+        result = subprocess.run(
+            [sys.executable, '-m', 'django', 'check', '--settings', 'postgres_tests.integration_settings'],
```
This fails when running from the root directory rather than the tests directory (i.e. `$ ./tests/runtests.py postgres_tests`). Also, the output doesn't make debugging very easy (`AssertionError: 1 != 0`) -- if that could be improved that could be nice.

timgraham commented on tests/postgres_tests/test_integration.py:
```diff
@@ -0,0 +1,21 @@
+from django.test import SimpleTestCase
+
+try:
+    import psycopg2
```
Did you consider using `PostgreSQLSimpleTestCase`? I would favor that for consistency.

jdufresne replied:
This test passes when testing against an SQLite or MySQL backend as long as psycopg2 is also installed. For this test case, I think the check `connection.vendor == 'postgresql'` skips the test too aggressively.

timgraham replied:
I know -- there are other tests on `contrib.postgres` that could run without a database, but we ran into other requirements like `django.contrib.postgres` must be in `INSTALLED_APPS` that made it inconvenient. I think it was 36e90d1f45a13f53ce25fdc2d9c04433b835c9af.

nasirhjafri replied:
@timgraham yes I tried with `PostgreSQLSimpleTestCase`, but `connection.vendor` is always returning me `sqlite` locally. So such tests won't run locally.

cansarigol replied:
you can use django-box, It is very useful to run tests with other vendors. (https://github.com/django/django-box)

nasirhjafri replied:
@cansarigol I can do it, but it would skip for other contributors locally. There is no use for a test which gets skipped.

charettes replied:
> I can do it, but it would skip for other contributors locally. There is no use for a test which gets skipped.

@nasirhjafri other contributors do run tests again other backends even if they are not using django-box.

nasirhjafri replied:
> > I can do it, but it would skip for other contributors locally. There is no use for a test which gets skipped.
> 
> @nasirhjafri other contributors do run tests again other backends even if they are not using django-box.

Hmm thanks, I've changed it to `PostgreSQLSimpleTestCase`

nasirhjafri commented:
buildbot test

timgraham commented on django/db/migrations/recorder.py:
```diff
@@ -18,23 +19,37 @@ class MigrationRecorder:
+    @classproperty
+    def Migration(cls):
+        """
+        When django.contrib.postgres in INSTALLED_APPS is in installed apps,
```
The note should be generic -- describe the sort problem as it could happen to any app -- a reference to contrib.postgres isn't needed.

timgraham commented on tests/postgres_tests/test_integration.py:
```diff
@@ -0,0 +1,16 @@
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL,
+        )
+        self.assertEqual(result.returncode, 0, msg="django.contrib.postgres integration test failed")
```
This isn't a helpful message -- tests show which test failed. What I meant was, if it's possible to display the check errors, that would be useful -- such as capturing stdout and/or stderror?
